main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1231,6 +1231,33 @@ object Typer extends Phase[ResolvedAst.Program, TypedAst.Root] {
           resultEff <- unifyEffM(evar, eff1, eff2, loc)
         } yield (resultTyp, resultEff)
 
+      case ResolvedAst.Expression.FixpointFold(pred, init, f, constraints, tvar, evar, loc) =>
+        //
+        // constraints : #{P : a | c}    init : b   f : a' -> b -> b
+        // where a' is the tuple reification of relation a
+        // ---------------------------------------------------
+        // fold P init f constraints : b
+        //
+        val freshPredicateTypeVar = Type.freshTypeVar()
+        val freshRestTypeVar = Type.freshTypeVar()
+        for {
+          (initType, eff1) <- visitExp(init)
+          (fType, eff2) <- visitExp(f)
+          (constraintsType, eff3) <- visitExp(constraints)
+          // constraints should have the form {pred.sym : freshPredicateTypeVar | freshRestTypeVar}
+          constraintsType2 <- unifyTypM(constraintsType, Type.SchemaExtend(pred.sym, freshPredicateTypeVar, freshRestTypeVar), loc)
+          tupleType = pred.sym match {
+            case rel: Symbol.RelSym =>
+              Type.mkTuple(program.relations(rel).attr.map(a => a.tpe))
+            case lat: Symbol.LatSym =>
+              /* TODO: what should be done in this case? */
+              ???
+          }
```

Comment:
What if we represent the type of a predicate as:

```
Apply(TypeConstructor.Rel(name), Apply(Apply(TypeConstructor.Tuple, t1), t2))
```

then we can write a constraint that requires the argument to fold to unify with the second component of the above type.

This would even work for the case when there is no tuple, but just one element?

Reply:
This looks reasonable indeed